forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt

```diff
@@ -0,0 +1,191 @@
+
+    override fun setAttribute(name: String, value: Any) = when (name) {
+        "userId" ->  userId = value as String
+        "challenge" -> challenge = value as String
+        "authenticationType" -> {
+            deviceBindingAuthenticationType = (value as? String)?.let {
+                DeviceBindingAuthenticationType.valueOf(it)
+            } ?: DeviceBindingAuthenticationType.NONE
+        }
+        "title" -> title = value as? String ?: ""
+        "subtitle" -> subtitle = value as? String ?: ""
+        "description" -> description = value as? String ?: ""
+        "timeout" -> timeout = value as? Int
+        else -> {}
+    }
+
+    override fun getType(): String {
+        return "DeviceBindingCallback"
+    }
+
+    fun setJws(value: String?) {
+        super.setValue(value, 0)
+    }
+
+    fun setDeviceName(value: String?) {
+        super.setValue(value, 1)
+    }
+
+    fun setClientError(value: String?) {
+        super.setValue(value, 2)
+    }
+
+    @JvmOverloads
+    fun bind(context: Context,
+             listener: FRListener<Void>,
+             keyAware: KeyAwareInterface = KeyAware(userId),
+             encryptedPreference: PreferenceUtil = PreferenceUtil(),
+             executorService: ScheduledExecutorService? = Executors.newSingleThreadScheduledExecutor(),
+             biometricUtil: BiometricUtil = BiometricUtil()
+    ) {
+
+        val keypair = createKeyPair(context, listener, keyAware) ?: return
+
+        if(deviceBindingAuthenticationType == DeviceBindingAuthenticationType.NONE) {
+            sign(context, listener, keypair, keyAware, encryptedPreference)
+            return
+        }
+
+        val biometricListener = object : BiometricAuthCompletionHandler {
+            override fun onSuccess(result: BiometricPrompt.AuthenticationResult?) {
+                if(isTimedOut) {
+                    setClientError("Timeout")
+                    Listener.onException(
+                        listener,
+                        DeviceBindBiometricException("Biometric Timeout Exception"))
+                } else {
+                    sign(context, listener, keypair, keyAware, encryptedPreference)
+                }
+                isTimedOut = false
+                executorService?.shutdown()
+            }
+
+            override fun onError(errorCode: Int, errorMessage: String?) {
+                setClientError("Timeout")
+                Logger.error(tag, errorMessage)
+                Listener.onException(
+                    listener,
+                    DeviceBindBiometricException("$errorCode: + $errorMessage"))
+                executorService?.shutdown()
+            }
+        }
+
+        biometricUtil.getBiometricAuth(biometricListener,
+            InitProvider.getCurrentActivityAsFragmentActivity(),
+            title,
+            subtitle,
+            deviceBindingAuthenticationType,
+            description)
+
+        if(isSupported(biometricUtil)) {
```

Comment:
Should check isSupported before generate keys?

Reply:
if the deviceBindingAuthenticationType == none , then we can generate the keys without biometric . i think we can modify the code little bit here.